applications/nrf5340_audio/src/modules/music_color_sync.c

```diff
@@ -0,0 +1,162 @@
+				frequency_modules[0] += (fft_data[i] << 7);
+			} else if (i < CONFIG_RGB_GREEN_FREQ_BIN_RANGE) {
+				frequency_modules[1] += (fft_data[i] << 7);
+			} else {
+				frequency_modules[2] += (fft_data[i] << 7);
+			}
+		}
+	}
+
+	/* Reducing intensity of less dominant modules*/
+	if ((frequency_modules[0] > frequency_modules[1]) &&
+	    (frequency_modules[0] > frequency_modules[2])) {
+		frequency_modules[1] *= 0.5;
+		frequency_modules[2] *= 0.5;
+	} else if ((frequency_modules[1] > frequency_modules[0]) &&
+		   (frequency_modules[1] > frequency_modules[2])) {
+		frequency_modules[0] *= 0.5;
+		frequency_modules[2] *= 0.5;
+	} else if ((frequency_modules[2] > frequency_modules[0]) &&
+		   frequency_modules[2] > frequency_modules[1]) {
+		frequency_modules[0] *= 0.5;
+		frequency_modules[1] *= 0.5;
+	}
+
+	/* Calculating duty cycle values, +1 added to prevent division by 0 */
+	uint32_t total_modules_value =
+		frequency_modules[0] + frequency_modules[1] + frequency_modules[2] + 1;
+	seq_values->channel_0 = 100 - (frequency_modules[0] * 100 / total_modules_value);
+	seq_values->channel_1 = 100 - (frequency_modules[1] * 100 / total_modules_value);
+	seq_values->channel_2 = 100 - (frequency_modules[2] * 100 / total_modules_value);
+}
+
+static int music_color_sync_thread(void *dummy1, void *dummy2, void *dummy3)
+{
+	arm_rfft_instance_q15 fft_instance;
+	arm_status status;
+	q15_t input[CONFIG_FFT_SAMPLE_SIZE];
+	/* Output has to be twice the size of input */
+	q15_t output[CONFIG_FFT_SAMPLE_SIZE * 2];
+
+	status = arm_rfft_init_q15(&fft_instance, CONFIG_FFT_SAMPLE_SIZE,
+				   CONFIG_FFT_FLAG_INVERSE_TRANSFORM, CONFIG_FFT_FLAG_BIT_REVERSAL);
+	if (status != ARM_MATH_SUCCESS) {
+		LOG_ERR("Failed to init FFT instance - (err: %d)", status);
+		return status;
+	}
+
+	while (1) {
+		k_sem_take(&music_color_sync_update_sem, K_FOREVER);
+		k_sem_take(&music_color_sync_value_sem, K_FOREVER);
+		memcpy(input, decoded_pcm_data_mono, CONFIG_FFT_SAMPLE_SIZE);
+		k_sem_give(&music_color_sync_value_sem);
+
+		arm_rfft_q15(&fft_instance, (q15_t *)decoded_pcm_data_mono, output);
+		arm_abs_q15(output, output, (CONFIG_FFT_SAMPLE_SIZE * 2));
+		music_color_sync_fft_analyse(output, CONFIG_FFT_SAMPLE_SIZE);
+		nrfx_pwm_simple_playback(&m_pwm, &seq, 1, NRFX_PWM_FLAG_LOOP);
+		k_msleep(125);
+
+		STACK_USAGE_PRINT("music_color_sync_thread", &music_color_sync_thread_data);
+	}
+}
+
+int music_color_sync_data_update(char *pcm_data_mono, uint16_t len)
+{
+	if (k_sem_take(&music_color_sync_value_sem, K_NO_WAIT)) {
+		memcpy(decoded_pcm_data_mono, pcm_data_mono, len);
+		k_sem_give(&music_color_sync_value_sem);
+		k_sem_give(&music_color_sync_update_sem);
+		return 0;
+	} else {
+		return -EBUSY;
+	}
+}
+
+int music_color_sync_init(void)
+{
+	int ret;
+	static nrfx_pwm_config_t const pwm_cfg = {
+		.output_pins = {
+			MUSIC_SYNC_RED_PIN,
+			MUSIC_SYNC_GREEN_PIN,
+			MUSIC_SYNC_BLUE_PIN,
+			NRFX_PWM_PIN_NOT_USED,
+		},
+		.irq_priority	= 6,
```

Comment:
NB: This is the IRQ priority and must not be mixed with the thread priority. What is the reason for 6?